<a href="https://colab.research.google.com/github/ricky-kiva/ml-machine-beginner/blob/main/m2_l5_pca_sklearn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **PCA (Principal component analysis)**

In [1]:
# import libraries
from sklearn import datasets
from sklearn.model_selection import train_test_split

Import dataset

In [2]:
iris = datasets.load_iris()
attribute, label = iris.data, iris.target

Split train & test set

In [3]:
# split by 80%
X_train, X_test, y_train, y_test = train_test_split(attribute, label, test_size=0.2, random_state=1)

Build Decision tree model, **without PCA**

In [4]:
from sklearn import tree

decision_tree = tree.DecisionTreeClassifier()

# train the model
first_model = decision_tree.fit(X_train, y_train)

# evaluate model accuracy
first_model.score(X_test, y_test)

0.9666666666666667

Applying **PCA** (Exploration)

In [6]:
from sklearn.decomposition import PCA

# instantiate PCA object with 4 principal component
pca = PCA(n_components=4)

# apply PCA to dataset
pca_attributes = pca.fit_transform(X_train)

# see variance of every attributes
pca.explained_variance_ratio_

array([0.92848323, 0.04764372, 0.01931005, 0.004563  ])

Applying **PCA**
*   Using 2 value of the best PCA (0.92 & 0.047)
*   Because 0.92 + 0.047 = 0.967 (near 100)

In [8]:
pca = PCA(n_components=2)
X_train_pca = pca.fit_transform(X_train)
X_test_pca = pca.fit_transform(X_test)

Train & score model **using PCA**

In [9]:
model2 = decision_tree.fit(X_train_pca, y_train)
model2.score(X_test_pca, y_test)

0.9333333333333333

**Result:** Model still have a high accuracy with reduced components